_posts/2020-08-05-hive-acid-original-files.md

```diff
@@ -0,0 +1,130 @@
+|                   |                    |    n_name : string,          |
+|                   |                    |    n_regionkey : int,        |
+|                   |                    |    n_comment : string        |
+|                   |                    |  }                           |
+|                   |                    |}                             |
+|                   |                    |                              |
+
+
+
+Before Presto 340, Presto used to fail the query if it reads from a  Hive transactional table having original files.
+
+# Update and delete support on original files
+
+Hive achieves updates/deletes on a row in original files by synthetically generating ACID columns for original files. Presto follows the same mechanism of generating ACID columns
+synthetically as discussed later.
+
+## ACID Columns’ generation on original files
+
+Files in Hive ACID format have 5 ACID columns, but we need only 3 columns i.e. `originalTransactionId`, `bucketId` and `rowId` to uniquely identify a row. In this section we will
+see how these 3 columns are synthetically generated for  original files.
+
+### Original transaction ID
+
+An original Transaction ID is the write ID when a record is first created. For original files, the original transaction ID is always the starting value i.e. 0.
+
+### Bucket ID
+
+Bucket ID is retrieved from the original file name. For the original file `0000ABC_DEF` or `0000ABC_DEF_copy_G`, the bucket ID will be `DEF`.
+
+### Row ID
+
+Row ID is calculated by taking the summation of row counts over all the files which come before the current original file when sorted in a lexicographical order belonging to the
+same bucket plus the local row ID in the current original file.
+
+Here is an example to calculate the global Row ID of the 3rd row of an original File `000000_0_copy_2`.
+~~~~
+000000_0            -> 	X1 Rows (returned by ORC footer field numberOfRows)
+
+000000_0_copy_1     -> 	X2 Rows (returned by ORC footer field numberOfRows)
+
+000000_0_copy_2     ->	[ Row 0 ]
+                        [ Row 1 ]
+                        [ Row 2 ]   <- Local Row ID (returned filePosition in OrcRecordReader) = 2, Global Row ID = (X1+X2+2)
+                        [ Row 3 ]
+
+000000_0_copy_3     ->  X4 Rows
+~~~~
+
+>**Note:** As we see, additional computations are required to generate rowIDs while reading original files, therefore, read will be slower than ACID format files in the transactional
+table.
+
+Once we have the 3 ACID columns for a row, we can check for update/delete on it. Delete deltas, written by Hive for Original files, have rowIDs generated by following the same
+strategy as discussed above, hence, the same logic of filtering out deleted rows as discussed in
+[Hive ACID and transactional tables' support in Presto](https://prestosql.io/blog/2020/06/01/hive-acid.html) works with the original files too.
+
+# Changes in Presto to support reading original files
+
```

Comment:
don't have two titles after each other without content in between

Reply:
yea, I have added some content here too.